### ISSUE TITLE
Multi-tenancy with custom domains

### ISSUE THREAD

**Issue by Gargron** [tootsuite/mastodon#2668]
More technical look at #897

Idea: Let users point DNS records to an existing instance, configure specific user to be served as-if from that domain.

What parts must work correctly for this:

- [ ] Separate field for storing whether an account is remote or local rather than relying on `domain` being null - store custom domain in that column?
- [ ] All functions that check whether an account is local must check the custom domain as well as the "local domain" of the instance
- [ ] Webfinger must answer queries for the faux-remote account
- [ ] *Other* accounts must not be served from that custom domain. Accessing the *website* through the custom domain should be limited to only public content from the owner of the domain.

Concerns:

- Switching instances while keeping the same domain will present the same problem as erasing/reinstalling an old instance on the same domain - other servers will remember data that you don't, any numeric IDs may reset and overlap (but generated string URIs include not just numeric ID, but also timestamp of creation, so that's not a big concern)
- Might be the first credible case for exporting/importing data *verbatim*, e.g. including all small details. Just like with local users having a custom domain, statuses should be amended to be local *and* able to have verbatim URIs like remote ones

**Comment by ZiiX** [comment]
Great option to consider! You know I'm for it. With this feature, I'm thinking it would be also awesome to consider personal file/media settings as an option with this. WebDAV, S3, Dropbox, Amazon, Google, etc - this allows their media hierarchy to follow them too.

**Comment by ramsey** [comment]
I feel like there are a lot of security/identity concerns to consider, especially with importing data verbatim. I could forge an import and pretend to be someone else, replaying all their posts into the new server. There’s no mechanism for signing posts and asserting one’s identity.

**Comment by saper** [comment]
@Gargron what about having a separate domain table and foreign key to that.

Also I think the primary domain should be treated just like all those tenants. Do you have a  reason to have it "special"?

**Comment by Gargron** [comment]
@saper Initial design decision was:

- Your domain could be adjusted (even just only at the very start when you're setting up/troubleshooting)
- Don't want to have 20.000 copies of the same string in the db if it can be avoided by making it null and generating local URIs on the fly
- Simple distinction for local vs remote content - remote content has a non-null URI

Of course right now we're exploring a use case where, perhaps, that design should be replaced... At the same time, you're right, we could have a separate domain table, then there wouldn't be 20.000 copies of the same string. Though you'd need to join the accounts and domains table on basically every single query.

**Comment by saper** [comment]
> Though you'd need to join the accounts and domains table on basically every single query.

This is something PostgreSQL should not have any problem with. Definitely we should not be storing
multiple copies.

**Comment by jeremyBanks** [comment]
Will it be possible to point to a different user name? If you need to use the same name across instances, that could limit your ability to migrate.

**Comment by ThibG** [comment]
Off the top of my head:
- Some of this is already handled with the distinction between `LOCAL_DOMAIN` and `WEB_DOMAIN`, except that you'd need to use a custom user-specific domain if not null instead of `LOCAL_DOMAIN`, which is easy when generating something, but can get messy when looking up an user. Also, Mastodon still makes some assumptions that are broken when `WEB_DOMAIN` is different from the user's domain (see #2012 and #2032), although until now, this can be worked around, but…
- What about username conflicts? Should user@redirected and user@local conflict or not? If not, the workaround for misdirected webfinger queries (https://github.com/y0t4/mastodon/commit/46555ce14df83391cdf222e574d170a7eebe7a8c) can't work in this setting, as it would need to disambiguate.
- Further, should the URIs of redirected accounts follow the same template as local ones? If so, and user@redirected and user@local are different users, how to disambiguate? The easiest way to handle that is probably to have https://web_domain/@user@redirected and https://web_domain/users/user@redirected etc.
- With that in place, using a remote Mastodon instance for your domain seems easy: just redirect (or proxy, but #2375) /.well-known/host-meta (and possibly /.well-known/webfinger), or use a CNAME RR for your domain, if it's to be only used for Mastodon.
- Migration probably will be a bit tricky: in OStatus, users are not identified by their acct: (user@redirected) URIs but by their ActivityStream actor URI (which is so far an URL to the user's web profile, and will thus reference the “backend” Mastodon instance). The easiest way to migrate seem to just point to another backend, however, your actor URI will most probably change as a result. I'm not sure how bad a thing that is. Also, it won't be immediate, as remote instances having your account details will need to refresh it from webfinger to switch to the new instance, and that's currently only done in a few situations, and at most once a day.

**Comment by semarie** [comment]
having this kind of configuration at DNS level would be great. But isn't more simple to have it at webserver level ? it would mean having a minimal code on remote webserver side (could be done with simple PHP script and/or webserver redirection).

As example, considere user@mastodon.social who want to use user@my.domain

On the remote side (my.domain):
- `/.well-known/host_meta` : reply with LRDD link to webfinger on my.domain
- `/.well-known/webfinger` : get information (mostly magic-public-key) from mastodon.social, and build a reply with acct:user@my.domain. Use local URL for atom, salmon, ...
- for URL entries points (atom, salmon, ...), use HTTP redirection (Location HTTP Header) to point to mastodon.social

On mastodon side (mastodon.social):
- permit the user to config is account as `user@my.domain`
- use `acct:user@my.domain` in replies

**Comment by russss** [comment]
I'm keen on @semarie's proposal, and in fact I started looking into the feasibility of it a while ago. The advantages as I see it are:

* You don't need a separate domain name purely for this - which means (for example) my username could be the same as my email address.
* Multiple users on the same domain could be housed on different Mastodon instances.

This would be similar to how OAuth works, and it feels like a more "federated" situation than pointing domain names around.

It also seems that in the original proposal, the host Mastodon instance would have to obtain an SSL certificate for each domain (or serve them unsecured which is annoying). Even with LetsEncrypt this is fiddly.

**Comment by saper** [comment]
here is what I understand under domain multi-tenancy:

Ability to point multiple A and AAAA DNS entries, add a new domain to the certificate and make sure the domains do not get mixed up.

1. Different domain names are treated like independent instances.

1a the simplest implementation would be just to reuse the code base and point to different databases on request. This is probably achievable now with a web server configuration. I think this should be tested prior to proceeding further (how to talk to redis, store media for multiple instances, mail support etc.).

Every instance is completely independent, has its own member namespace and they federate independently. They only get upgraded at the same time. You have to run db:migrate multiple times, with a new RUBY_ENV for ever domain in the loop.

I think there must be some folk trying this right now?

MediaWiki has an interesting hack that allows sharing only some database tables between such instances, not sure it would be helpful here.

A dedicated asset domain could be another thing, but it should be made in a way that does not break the possibility to export the domain to a fully standalone instance.

1b accept multiple WEB_DOMAINS and issues and hacks related to that. This is more tricky in the implementation as it might require a new `domain` table, new queries etc. This provides little more benefit to 1a but it carries more of the development effort. The web server/PostgreSQL configuration is simplified compared to (1a)and every domain lives independently in the webspace, with its own member namespace.

Interesting questions appear here, like should those instances automatically be federated or not? 
(I think not, they should feel independent).  

There is a possibility to shortcut the traffic between the domains internally for performance reasons (maybe not necessary).

1c Accept multiple domains under one certificate - basically it is the scheme used by Google multi-tenant Apps. Everything is under google.com/a/domainname/restoftheurl 

The problem here is the initial redirect - how to securely redirect domain.fluffy to mastodon.social/a/domain.fluffy on the web? In case of SMTP it works because nobody validates SMTP domains in certs and we have MX records.

One solution could be to have some custom SRV record that is universally accepted (haven't checked the spec yet). StatusNet, bot frameworks etc should also use it.

A quick check with the WebFinger RFC tells me it only supports HTTP redirect to a different domain and not a well known URL but you still need to be able to accept initial request securely.

And here comes another variant:

2a. Some fluffy form of "user@domain" as an identity alias on mastodon.social without DNS entries. I don't know how would that possibly work, except for the user being `mrserious@mastodn.social`vall the time on the protocol level and just bring presented as `@pinky@fluffyanimal.farm` on mastodon.social only. 

2b. A variant of (2a) could be considered where mastodon.social tries to convince other instances that `mrserious` wants to be presented as `@pinky@fluffyanimal.farm`, which opens a whole can of worms (misrepresentation).

May be this use case should be further elaborated (maybe it is a separate issue).

I think there is a natural development progression to go from 1a via 1b maybe to 1c.

it would be good to refer to one of those variants as we speak (or define additional ones)

**Comment by s-leroux** [comment]
Great idea. But let me rephrase to see if I understand the typical use case:

Someone connects to me as @sylvain@yesik.it

First we query the yesik.it DNS for the _mastodon TXT record.
If not found, "yesik.it" is assumed to run its own Mastodon instance.
Otherwise, the TXT record (CNAME?) holds my current instance. Say its "mastodon.social"

So, on the background, Mastodon will understand the _physical_ account name is @sylvain@mastodon.social
But in the UI and for all user interactions, only my _logical_ account name is used @sylvain@yesik.it

Basically we add here a new layer of indirection. Or did I missed the point?

----

Several remarks:

1. It doesn't help if someone bookmarked my account URL (https://mastodon.social/@YesIKnowIT). Or does it?
2. What if several members of the _same_ domain have accounts on _different_ instances?
3. What if I actually made _several_ domains point toward the _same_ account (aka, @sylvain@chicoree.fr, @sylvain@yesik.it ... being different alias to the same account -- just like I can do it for my email).

**Comment by saper** [comment]
@s-leroux so what you mean is something like federated alias.

Basically one more assumption had to be broken - not all users from @domain are hosted on a single instance. 

That is another piece of complexity, basically "domain" server would need to provide redirects/pointers,
but I understand that the trouble is for the target host to accept user@otherdomain stuff and somehow treat it's as own

Not sure it's multi-tenant domain handling, though

**Comment by alexgleason** [comment]
I like this idea a lot, but not sure the benefits outweigh the costs.

I think the goal is to give users more control. The easiest way to do that now is to run your own instance. That's hard though for non-tech users, while setting some DNS values is easy given a tutorial.

Would it be better to make it easy to start your own instance? I feel like a 1-click install on a web host would be ideal. I think IndieHosters is supporting Mastodon, maybe they'd be willing to set something up like this. @pierreozoux

Basically I'm wondering what the goals actually are and how to best reach those goals. I can definitely see the benefits of supporting remote domain names, but running your own instance is more beneficial: you own/control your data, there's no reason to migrate, and it produces no technical debt to the project itself. The paradigm currently in place makes sense: if you're part of a server, you're part of that community and you take their name. If you're remote, you have a remote name. This is intuitive.

On the other hand, email has some similar paradigms and any service can send emails from your domain name with the proper authentication (eg Postmark). But this is usually a generic service and to my knowledge is not built into any email providers themselves.

**Comment by pierreozoux** [comment]
I think it is a tricky discussion.

In one hand, with one postfix, I can host various domains, and it is convenient.
As a user, I can migrate from one hoster to another, it is a bit painful, but it is doable.

On the other hand, on the web, one domain name means the control over your own piece of land, your rules, your design, so it also makes sense to have your own instance (from the indieweb principles).

It is not that hard to host a mastodon instance, or pay somebody to do it. And then, it is easy to just transfer from one server to another, or from one hoster to another. And the change that has to be made to add this feature are quiet heavy I guess. Then, they also have to be maintained.

I think that multitenancy, is always a headache for developers, and it is never really well done, so even if I have the option, as a hoster, I never choose that. It is always easier to migrate people with their own instances.
(All the options I hear, DNS, running a small app server will touch 0.01% of our user base IMHO)

I'm probably not the best judge to help here, but I think the discussion has to evolve around cost/benefit for the software in itself. I'm a fan of unix philosophy, and I think the effort can serve a lot better in other areas :)

And to conclude, I also think that the discussion should happen at the standard itsefl, at ostatus level.

**Comment by jeremyBanks** [comment]
"It's not hard to host a Mastodon instance", said one engineer to the next.

Dozens of projects have told themselves this, but the reality of self-hosting applications like this is that *it does not happen* for a non-trivial number of users, despite great efforts like Sandstorm. One-click install for end-users is not a responsible solution, unless you also have automatic fast and **secure** updating on multiple levels. We might get there one day, but don't let that ideal get in the way of a realistic solution.

Configuring DNS entries as a one-time (or at least rare) action is something that a moderately tech-competent person can handle. It will at least get Mastodon on par with email in terms of identifier portability, and is probably the most viable mid-term solution (assuming the current identifier scheme is preserved instead of moving to something cryptographic).

**Comment by DanielHeath** [comment]
Postfix is a good example of something that is easy to operate. 
Enable unattended-upgrades and security patches happen magically.
You can automatically transfer all your mail from one provider to another via standard APIs.
This is still not enough - the vast majority of users prefer free-with-ads (eg gmail) or pay-for-hosting (eg fastmail).

Hosting your own Mastodon instance means caring about:

 * Backups (you presumably care about your own history)
 * Patching security bugs quickly (Much as I love rails, it requires more frequent patches than other stuff I run)
 * SSL (letsencrypt has made this easier, but it's still a job to setup)
 * Diagnosing & fixing issues (if something goes wrong, can you pull the relevant bit of the logs to file a bug?)
 * Uptime (ok, it's a personal blog, but it still needs to be fixed when it crashes)

I build and operate production services for work - I don't need or want another thing to carry pager for.

**Comment by DanielHeath** [comment]
@s-leroux I think a mastodon-specific TXT record is unnecessary; a regular old CNAME should do fine.

I think that separate-but-related is the ability to redirect an account from an old name to a new name. It's nice to be able to change your identity (even on the same domain) without breaking all the old URLs and losing your federated followers - there's no reason this sort of redirect shouldn't work across domains too.

**Comment by s-leroux** [comment]
@DanielHeath The problem with CNAME is cannot coexist with _any other record_ for the same resource (except DNSSEC-related records).

In particular, you cannot have both a CNAME and a MX record for the same domain.
Practically, if you use a CNAME to point your bare domain to some "mastodon provider", you no longer are able to use a mail provider on your bare domain. The workaround being to create a sub-domain per service. But that feels ugly to me.

**Comment by DanielHeath** [comment]
You can use an A record with an MX record. 

That gives you mastodon and email, but you'll have to use other domains for any other protocols. 

If you need something else, you can run nginx as a reverse proxy, or host your own. 

Thanks,
Daniel Heath

**Comment by s-leroux** [comment]
@DanielHeath I understand your point of view concerning someone having full access to his server.

But imagine a business or other kind of organization deciding to provide Mastodon access to their team.
 
However, they already have mail and web site on their primary naked domain (MX and A records). In most cases those services are externalized. The web server having chances of being a mutualized one.
The only standard record that _might_ "bind" their domain to some mastodon provider would be the SRV record. Unless we force the user to create a specific sub-domain? Anyway, in the cases I have in mind, if I explain to them they will now have to host themselves a server just to install a reverse proxy or some any other service -- surely they will answer, "never mind, we will continue using TweetDeck".


My point is if we want to target a wide audience, any viable solution should satisfy those two constraints:
1. It should be possible to co-host Mastodon as well as classical email and web services on the primary naked domain
2. For the purpose of adding the "Mastodon alias", we should assume the user only has access to his domain DNS configuration.



Re-reading the entire discussion, I cannot resist quoting @saper :

> Not sure it's multi-tenant domain handling, though

**Comment by DanielHeath** [comment]
My issue there is that it breaks ostatus protocol compatibility to achieve something that can be implemented without breaking compatibility.

If the project didn't need to adhere to existing federation standards there are lots of nice features that could be added; I don't think that's what mastodon is about though.

Thanks,
Daniel Heath

**Comment by duck57** [comment]
One other use case for this (assuming I understand correctly) is for a current self-hosting user to add other domains to their existing instance.  This probably will mostly be used for novelty accounts, but having the ability to use one instance of Mastodon on your server to serve several domains would be a nice-to-have feature and there are probably better uses than silly account domain names that I'm just not thinking of at the moment.

**Comment by azbulutlu** [comment]
some questions, as a user.

1) I use shared hosting, I have my domain, that has my website and email address. Can I use this feature? (not everyone can afford to pay for additional vps hosting. not everyone has time/skill/energy to run a server. even if a person has a vps, running a public facing server with good uptime is a lot of overhead on top of that)

2) What is my profile url? (note that when giving social media links on a website, I use the profile url).

3) what is my mastodon handle? (if I give someone my profile link, ideally they should be able to add me back, like they do with the profile url of a regular account. So follow button, or typing @username@customdomain or pasting the link and finding the account that way)

4) (potentially silly question) in terms of the idea for shared certificate, my domain already has a certificate, does that mean I cannot use this domain? 

what are the answers in case I use a subdomain that is dedicated to mastodon redirecting instead (e.g. social.mycustomdomain.com)

@duck57 A scenario would be that: you can have 2 domains (e.g. both your personal page and a project/blog) You might want to have "official" accounts for both, without having 2 mastodon installs  (which would be a big overhead). It is not uncommon to have separate social media accounts for a personal project vs personal page, even if the project is not that of a group.

**Comment by duck57** [comment]
@azbulutlu Thanks for coming up with a much more "professional" use case than I thought of.  One install per server makes much more sense than one install per domain.  Perhaps you could use some NGINX magic to proxy traffic from one domain to the other as an interim solution, but someone with more NGINX experience than I should comment if this is actually feasible as a solution or not.

**Comment by DanielHeath** [comment]
@duck57 trying to proxy is fraught - you have to correctly rewrite *every* instance of the domain name in the body of all resources, but you also have to consider whether you're breaking CSP (since you might have changed the body of a resource which has a hash provided).

It can be made to work but it's not something I'd want to rely on, especially if I'll be updating the backend over time (which might introduce new ways for the nginx config to be wrong).

**Comment by duck57** [comment]
@DanielHeath Something I have seen people do is use cnames, but I'm not sure if that solves this issue or simply allows a user to access multiple accounts on a single instance through an app that assumes one account per instance per user.

**Comment by raphaeljolivet** [comment]
There are plenty of geeks out there running their own servers and not mutualizing things. I guess this impacts badly the performances of the fediverse as a whole :  shared inboxes can't be used and popular toots are replicated to many single-user instances.

To me the [solution of](https://github.com/tootsuite/mastodon/issues/2668#issuecomment-298299007) @semarie is the way to go.
So let's go.

The first step would be to add a field **customDomain** in the account table, and spot the places where the username @Foo is assumed to be @Foo@instance-domain.fr 

For instance the user profile page should accept a full handle like :
* http://instance-domain/users/Raph@my.domain.com 
* http://instance-domain/@Raph@my.domain.com

Once the WebFinger resolution is done, I guess **ActivityPub** only refer to full URLs for user IDs, so everything should be fine. right ?

Next step should be to support this in user migration. Typically : migration from the default domain to the custom one, possibly on the same instance. How would migration be handled when moving from one provider to another, with the same handle ?

I would be interested  to have the point of view of an active developer on this code base, so we could spot limitations and potential issues with this approach.

**Comment by DanielHeath** [comment]
@semarie I *really* appreciate the concept of splitting webfinger out from hosting - that's a really nice design that avoids quite a bit of complexity.

**Comment by raphaeljolivet** [comment]
So ... I have managed use use my custom domain and make it point to a shared instance, without hacking Mastodon 
... and, it kinda works ...

I have developed a small python app  [webfinger-proxy](https://github.com/raphaeljolivet/webfinger-proxy) to alias your own domain handles to a shared instance. 

I have setup it on my own domain [rjo.name](https://rjo.name).
The handle `testrjo@rjo.name` points to `testrjo@mamot.fr` 
See the response of the Webfinger [https://rjo.name/.well-known/webfinger?resource=testrjo@rjo.name](https://rjo.name/.well-known/webfinger?resource=testrjo@rjo.name)

I have managed to follow this account and send it private messages.
I have another account on the same shared instance, and `testrjo@mamot.fr` appears as a completely separate remote account though. 

I am not sure whether this is only on the UI or if Mastodon now have a duplicate account for this guy ... 
I would need some help from the developers to figure this out : Is the handle a primary key ? Or is it the URL ?
Can you see any have adverse effects ?
Do you think the physical migration of accounts from one instance to another would still work ?

I will continue my tests with other implementations : 
Pleroma, mobile clients, etc ....

**Comment by almet** [comment]
Thanks to @raphaeljolivet last message, I took a similar approach, but with an nginx configuration and static files, since I didn't wanted to have a service running just for this.

If you want to do the same, [here's a link to my work](https://github.com/almet/.well-known).

**Comment by almet** [comment]
This seems to have limitations currently, since there are now ways to define in the settings what our "main" identity is.

I have two accounts : the "front" one which I want others to see and the "back" one. If I interact with users (answer, follow, etc), I should be seen with my front identity, and the back one shouldn't appear anywhere, I don't think that's what's happening here.

**Comment by ThibG** [comment]
Please do not do that, and wait for solutions to be implemented in Mastodon itself, since it will require cooperation with the hosting instance.

**Comment by raphaeljolivet** [comment]
> Please do not do that, and wait for solutions to be implemented in Mastodon itself, since it will require cooperation with the hosting instance.

Ok fibne, I have added a warning note to my repo :
https://github.com/raphaeljolivet/webfinger-proxy/blob/master/README.md

**Comment by Exagone313** [comment]
Personally I started to follow this issue because I'd like to open my instance to new users.

My Mastodon instance uses a subdomain that I created as a temporary handle, in 2017, that doesn't use a domain name I'd like to give to others.

I don't want to break my own account and transfer to another domain name completely, I just want to be able to add a new domain name to the server that could be used by new users. A bit like an email provider can allow users to register using different domain names.

I feel like this current issue is not what I am looking for. There are different needs between a user that wants to use their own domain name on a third party Mastodon instance and a Mastodon instance administrator that wants to manage multiple domains.

(Also I'm against the idea of allowing multiple users of the same domain name to use different Mastodon instances. And I understand there would be issues to move a domain name between instances.)

**Comment by Saklad5** [comment]
> I don't want to break my own account and transfer to another domain name completely, I just want to be able to add a new domain name to the server that could be used by new users. A bit like an email provider can allow users to register using different domain names.

That's made possible using MX records, which separate identity (the email domain) and location (the IP addresses of the SMTP servers). SVCB records would do the same for Mastodon.

**Comment by vincentbernat** [comment]
No need to wait for SVCB (only a draft, then it needs to be supported by DNS servers, this will take quite some time). Mastodon can make use of SRV records instead (which are a generalization of the MX records).

**Comment by Saklad5** [comment]
> No need to wait for SVCB (only a draft, then it needs to be supported by DNS servers, this will take quite some time). Mastodon can make use of SRV records instead (which are a generalization of the MX records).

They're an Internet Draft, yes, but they already have widespread deployment and finalization is imminent. Besides, DNS servers **don't** need to support every record they handle.

The HTTPS record type is designed to be extended by protocols that use HTTP: I would **strongly** encourage drafting a specification that does so.

```DNS
_mastodon.example.com.		300	IN	HTTPS	1 actual-mastodon-instance.example alpn="h3,h2"
```

That example could indicate that the Mastodon instance of `example.com` is located at `actual-mastodon-instance.example`, which accepts HTTP/3, HTTP/2, and HTTP/1.1. When Encrypted Client Hello is finalized, Mastodon records could adopt it as well.

**Comment by vincentbernat** [comment]
Many DNS servers don't accept arbitrary records, notably the ones from cloud providers. Mastodon should at least support a fallback method in this case (SRV or well-known location).

**Comment by Saklad5** [comment]
> Many DNS servers don't accept arbitrary records, notably the ones from cloud providers. Mastodon should at least support a fallback method in this case (SRV or well-known location).

Those DNS nameservers will have to support HTTPS records if they intend to allow HTTP/3: Alt-Svc headers are being deprecated in favor of HTTPS records, which will make them mandatory for that protocol.

If a DNS nameserver insists on imposing arbitrary restrictions on DNS zones, people should stop using them. Heck, self-host the nameserver! DNS is designed with massive amounts of caching for a reason. DigitalOcean is a notable offender, but at the same time they literally have [a guide for running your own BIND server on their infrastructure](https://www.digitalocean.com/community/tutorials/how-to-configure-bind-as-a-private-network-dns-server-on-ubuntu-14-04).

**Comment by steelman** [comment]
It's mostly not a problem of DNS servers but rather those web cpanel hosting users get in their bundle. And there is a lot of them, and service providers aren quick to update them.

**Comment by Saklad5** [comment]
It is an extremely bad idea to cater to the lowest common denominator.

**Comment by untitaker** [comment]
Imagine you're a content creator who has their blog and shopify on one domain, and now Mastodon requires you to potentially disrupt that setup just to be able to _add_ it as a potential revenue stream. Are you going to ask those people to self-host their DNS servers or move hosting providers?

**Comment by Saklad5** [comment]
Mastodon doesn't have to be SVCB-reliant. In fact, given the number of implementations that already exist, it couldn't be for a fairly long time. If someone can't use SVCB records, they can just use the fallback SVCB-optional behavior. That is, after all, the _only_ behavior that actually exists right now.

**Comment by irrelevantdotcom** [comment]
To add a usage case, and point of view.. I've just set up an instance for personal (family and friends) use. This is likely to be fairly lightly used (as in, in the short term, it'll be just me!)  I'd also like to set up an instance for a particular interest group I'm a member of. This might attract a dozen or so users. These would not naturally want to share the same domain name, nor would I want them to. Given the overhead per instance, it would make sense in my eyes for them to be able to share one server, much as I load up one webserver with several small sites.

**Comment by duaneking** [comment]
The instance should honestly just accept a list of domains it should respond to, blocking input from others that do not match; Things like CORS make the need for this list explicit and managed on the server side, anyway.

And then you have email outbound; Email outbound about account messaging for every domain used, so are you using a central email account for all the TLD's of that deployment, or a specific one fir each of the supported domains?  What about DKIM? Is the goal to using be using SPF text records and abandoning DKIM security?

**Comment by manedfolf** [comment]
Since it got brought up. I would still love to see something like this happen. Either users utilizing their own domains or instances truly supporting multiple domains on the front end as described in #8921

**Comment by rfc1036** [comment]
You can locally implement WebFinger on your own domain (here I have a proper recipe for Apache: https://blog.bofh.it/debian/id_464), but while this allows other to look up that identity it does not change how you are known to the Fediverse.

Also, there is software around which takes shortcuts and does not properly implement the WebFinger specification (e.g. https://github.com/lucahammer/fedifinder/issues/166 https://github.com/lucahammer/fedifinder/issues/171).

**Comment by ClearlyClaire** [comment]
> Mastodon uses WebFinger to look up user@domain's actor URL. Can't the user set up their WebFinger server to return the Mastodon's URL and thus retain control of their identity? This would do away with any changes to be made in DNS.

That allows you to have an alias, but Mastodon needs a canonical identity, and that one is figured out by looping back with a webfinger request to the domain hosting the ActivityPub actor itself. If we used the first known alias as the actual handle, anybody could poison the database by making arbitrary aliases to your account.

Now, Mastodon *could* redirect back to the external domain, but that requires knowledge that that specific account is using another domain, and it's not something Mastodon is currently built for.

And that is for the multi-tenancy part itself, not for moving a Webfinger handle to another Mastodon/ActivityPub server. That is another issue: technically, as far as ActivityPub itself is concerned, Webfinger has no relevance, and the identity *is* the actor id, which in practice is its URL. So while Mastodon may consider the Webfinger handle as the primary identifier… that's not what the spec says, and other projects may do otherwise.

> This does assume that the entire fediverse follows the same lookup pattern. Servers which don't do the same lookup will cache the actor URL and ping the old server, even though user's WebFinger response has started pointing elsewhere.

Yes. A bunch of implementations actually don't use Webfinger in the same way as Mastodon does, with it being only a one-directional discovery thing.

**Comment by zemlanin** [comment]
> As an aside, I actually think WebFinger should be made easier for static hosting

@nileshtrivedi You can achieve this now, by serving `/.well-known/host-meta.xml` with something like this:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<XRD xmlns="http://docs.oasis-open.org/ns/xri/xrd-1.0">
  <Link rel="lrdd" template="https://example.com/.well-known/webfinger/{uri}.json"/>
</XRD>
```

and hosting `/.well-known/webfinger/acct:name@example.com.json`. But a lot of fediverse clients don't check `/.well-known/host-meta.xml` before doing the WebFinger request :/

**Comment by nileshtrivedi** [comment]
@zemlanin Whoa! That's 🤯 

Is this part of the WebFinger spec? Does Mastodon support this?

**Comment by Saklad5** [comment]
It sounds like the main problem with WebFinger is the lack of a TTL. At any rate, WebFinger complaints should be directed to the W3C, and are out-of-scope for Mastodon.

In other words, if you want to fix a WebFinger-related Mastodon problem, you should be talking to them first. It'll also help all the other things using WebFinger.

**Comment by voxpelli** [comment]
> @zemlanin Whoa! That's 🤯
> 
> Is this part of the WebFinger spec? Does Mastodon support this?

Host-Meta is it’s own spec that was constructed to enable eg WebFinger: https://www.rfc-editor.org/rfc/rfc6415.html

Then WebFinger was kind of “hijacked” and “simplified” to no longer rely on LRDD / JRD and ended up with the current spec that does not work well with static sites.

So: No, it’s not part of the current WebFinger spec, but it was part of the original WebFinger proposal by Blaine Cook.

**Comment by zemlanin** [comment]
But Mastodon _does_ support it https://docs.joinmastodon.org/dev/routes/#host-meta

**Comment by Saklad5** [comment]
> Then WebFinger was kind of “hijacked” and “simplified” to no longer rely on LRDD / JRD and ended up with the current spec that does not work well with static sites.

I'm a firm believer in following specifications. If you aren't following the specification, call it something besides WebFinger. And don't rely on others following suit.

**Comment by Saklad5** [comment]
Wait, there **is** a system for changing providers:

> A WebFinger resource can include cache validators in a response to
> enable conditional requests by the client and/or expiration times as
> per [Section 13 of RFC 2616](https://www.rfc-editor.org/rfc/rfc2616#section-13).

Putting aside the obsoleted spec it references, WebFinger clients have to respect HTTP's `Cache-Control` parameters. Looking something up once and never again isn't really compliant.

**Comment by ThePowerOfDreams** [comment]
>WebFinger clients have to respect HTTP's Cache-Control parameters.

What makes you say that? They _should_, but does any spec or standard mandate that?

**Comment by Saklad5** [comment]
> > WebFinger clients have to respect HTTP's Cache-Control parameters.
> 
> What makes you say that? They _should_, but does any spec or standard mandate that?

Yes, [RFC 9111](https://www.rfc-editor.org/rfc/rfc9111). Since WebFinger is using HTTP, and explicitly references HTTP cache mechanisms, they are in force. As such, a WebFinger response can't be reused except as allowed by `Cache-Control`.

This doesn't stop applications from allowing a WebFinger request to only be made once, to be clear: it merely stops them from assuming a response is still authoritative for an origin outside those rules. If a spec says something is resolved using a WebFinger request, that means each resolution is a distinct WebFinger request, and therefore has to use those caching rules. If it says something is resolved using a database with initial entries populated by a WebFinger request, that means only the first resolution is a WebFinger request, and caching does not necessarily apply.

In Mastodon's case, a lookup is [defined](https://docs.joinmastodon.org/spec/webfinger/#intro) as being a WebFinger request, which means it should be following caching rules for each lookup.

Note that the lack of a `Cache-Control` header on certain HTTP status codes means the client can basically cache [at their discretion](https://www.rfc-editor.org/rfc/rfc9111#section-4.2.2).

**Comment by steelman** [comment]
Dnia 8 grudnia 2022 10:36:07 UTC, Nilesh ***@***.***> napisał/a:
>@rfc1036 Interesting.
>
>I actually think WebFinger should be made easier for static hosting. 

It's easy with two lines added to .htaccess https://gist.github.com/aaronpk/5846789

-- 
Łukasz Stelmach z podróży

**Comment by nileshtrivedi** [comment]
> Dnia 8 grudnia 2022 10:36:07 UTC, Nilesh ***@***.***> napisał/a:
> @rfc1036 Interesting. I actually think WebFinger should be made easier for static hosting.
> It's easy with two lines added to .htaccess https://gist.github.com/aaronpk/5846789
> […](#)
> -- Łukasz Stelmach z podróży

Static hosting environments come in a huge variety now a days. For eg: how do I apply your suggestion if my static site is on GitHub Pages? Cloudflare Pages? Netlify? Vercel? Firebase? Hacks for one webserver are not a replacement for a flawed protocol spec.

**Comment by jippi** [comment]
I've been slowly building out this Open Source guide generator for using your own domain - both static sites and other tooling :) 

https://guide.toot.as/guide/use-your-own-domain/

**Comment by willnorris** [comment]
Just regarding the discussion of caching (which I know is continuing to veer a bit of topic), the webfinger document itself includes an expiration field, though it doesn't seem mastodon uses it.  From the [example in RFC 6415](https://www.rfc-editor.org/rfc/rfc6415.html#appendix-A):

``` json
{
      "subject":"http://blog.example.com/article/id/314",
      "expires":"2010-01-30T09:30:00Z",

      "...": "..."
}
```

**Comment by filipesaraiva** [comment]
Just a doubt, is this feature in roadmap?

**Comment by dogshapedangel** [comment]
I am also curious on whether this will see any movement given Bluesky's custom domain handle functionality and the apparent demand for a feature like this. I've gotten complaints for years about how folks are concerned about my unchangeable domain name "yiff.life" and while I still think it's funny, it would be very cool to allow people to actually use our alternate aliased SFW domains for their handles, even if it is still just masking yiff.life. 

But now the idea may be a bit different with the advent of what Bluesky implemented. I still prefer the original multi-tenant idea.

**Comment by flamingsnowflake** [comment]
as tempting as it is, I don't think it's a good idea to compare Mastodon to Bluesky or Discord

those are user problems, not server problems

I totally get that we need three or more accounts per person and people want fancy URLs.
but it's a lot more important to separate those from each other... not everybody is cool enough to post pony porn or luigi memes with the same account that they use to apply for a job.

some people want their private life to remain private. If you run the server, you will be responsible for this. do you want this responsibility? not for you, for your users. 

maybe it's time to think about single user Mastodon instances with multiple domain names and multiple identities? 
then the isolation is no longer a problem. because it is the same user and the same database on the same machine but it only has to look different to the outside world. 

I'm not a developer but these are the things that prevent me from using Mastodon. I don't want to run three instances for one person with multiple identities. 

I also don't want to become a member on somebody else's server.

we are all spoiled by Discord and Bluesky.

**Comment by saschanaz** [comment]
> maybe it's time to think about single user Mastodon instances with multiple domain names and multiple identities? then the isolation is no longer a problem. because it is the same user and the same database on the same machine but it only has to look different to the outside world.
> 
> I'm not a developer but these are the things that prevent me from using Mastodon. I don't want to run three instances for one person with multiple identities.
> 
> I also don't want to become a member on somebody else's server.
> 
> we are all spoiled by Discord and Bluesky.

Perhaps you are looking for [Hollo](https://docs.hollo.social/) than Mastodon in that case.

**Comment by dogshapedangel** [comment]
> as tempting as it is, I don't think it's a good idea to compare Mastodon to Bluesky or Discord

I am not sure where DIscord comes into this, but Bluesky is another decentralized social media platform that is at least somewhat similar to Mastodon and ActivityPub. While I would not advocate Mastodon directly replicate what Bluesky is doing, this feature request is older than Bluesky's creation, and I absolutely think it is worth noting when another platform implements an extremely similar idea with great success. 

There are a few ways this would be beneficial to users:

- Alternate domains offered by admins 
- User domains for vanity URLs

I could also see it potentially causing confusion depending on how it is implemented. 

My inquiry is mostly just to understand whether this is still on the roadmap in some form or another, and it looks like some work was done in 2024 (https://github.com/mastodon/mastodon/pull/32724) to move away from using usernames as immutable IDs and instead use numeric generated IDs, which would be a huge step in the direction of supporting this.

**Comment by nileshtrivedi** [comment]
> Bluesky is another decentralized social media platform

It looks centralized to me. Most users are on the central server. This has been [called out by others](https://dustycloud.org/blog/how-decentralized-is-bluesky/) as well.

**Comment by dogshapedangel** [comment]
> > Bluesky is another decentralized social media platform
> 
> It looks centralized to me. Most users are on the central server. This has been [called out by others](https://dustycloud.org/blog/how-decentralized-is-bluesky/) as well.

I agree with your correction here but I do not believe that detracts from my belief that the feature would see high demand in both Bluesky's form (as a verified handle) and as originally proposed in this issue in multi-tenancy format.

**Comment by Exagone313** [comment]
I had a quick look at how Bluesky/atproto is dealing with handles. If you add a TXT record to a `_atproto` subdomain or if you configure your HTTPS server for your domain name to answer `/.well-known/atproto-did`, you can return a *did:plc* identifier. This identifier is then queried on https://plc.directory (then comes some cryptography stuff).

With ActivityPub (actually, WebFinger), you need to configure your HTTPS server for your domain to answer `/.well-known/webfinger?resource=acct:username@yourdomain.example` (for identifying accounts, to keep it simple).

So, to summarize:
* with atproto, you rely on a centralized entity (plc.directory)
* with ActivityPub, each domain is independent from each other, there is not a centralized entity

For Mastodon to support multiple domains, it needs to handle WebFinger (and other stuff) on multiple origins. There are multiple ways to consider this.

**Comment by corobin** [comment]
I am one of those people who use BlueSky's custom handles and found this thread because I was looking for the same with Mastodon. 

I just want to chime in with a use-case:

The BlueSky custom handle is very convenient because it makes my domain a universal identifier that does not depend on the instance's domain. it's also very easy to set up just with verifying the domain ownership. there's no need for me to spin up my own web server or run my own instance of the platform just to use my domain as a handle. lots of people would not want to run their own for a variety of reasons, such as limited technical ability, not wanting to spend the maintenance time or effort, or monetary cost of running a web server, but a DNS record is relatively trivial to set up and does not incur any marginal cost to do.

The BlueSky custom handle is also a bit different at the front end than - say - running your own instance. Your BlueSky username is simply `example.com` whereas with your own instance it would currently be something like `myname@example.com` or `me@example.com`. the user-part is just unnecessary for many many use cases, particularly where it's a single-person personal domain.

another benefit for the way BlueSky does it is that it makes it very pretty to set my domain as my handle, and also setup a redirect from my domain to my profile. it's very elegant for anyone who doesn't care to have a separate landing page. you can go directly to `example.com` to get to my bluesky profile, whose handle is also `@example.com`. without the redirect even though my handle is `@example.com` you won't actually get to my profile if you just typed in `example.com`. this kind of symmetry becomes ugly if you have to have an additional username before the domain part.


to sum up: at the very basic level, supporting bluesky-style domain handles would be a very useful feature for people who have their own domains and want that persistent identity, but for *whatever* reason want to use an established server (which is a lot of people).

**Comment by Exagone313** [comment]
Disclaimer: I have not read the full specifications, I am only writing this after quick reading and research.

---

Mastodon and other software implement existing specifications (like WebFinger). Adding new ways to identify accounts would not be supported by other software. I think it is too late to change these choices, or it could fragment the ecosystem.

To move an account to another instance, you would need to download the secret key for your account, upload it to the new instance and then the new instance would need to rotate the key for security reasons. From my understanding, if the instance disappears or bans you, or if you loose the secret key, your existing account handle won't be usable anymore, unless existing instances forget you.

It is not as easy as changing a DNS record.

---

I think that adding the support of multiple domains for accounts in a single Mastodon instance is a first step and the main feature request discussed here. Adding the ability to move a domain to another instance could only come later.

**Comment by andypiper** [comment]
The discussion of DID vs WebFinger, ATProto vs / and-or ActivityPub etc goes significantly outside of this issue. ActivityPub account portability is under discussion in working groups at the W3C. You'll need to explore other communities (W3C, activitypub.rocks, etc) to go deeper into those conversations as this issue isn't the right place to do so. 

There's no current/immediate plan to implement multi-tenancy with custom domains in Mastodon itself (and it is somewhat complicated by [the need to remove support for Redis namespaces in an upcoming version](https://github.com/mastodon/mastodon/discussions/34198)). We'll certainly keep the issue open as a suggestion for now.

**Comment by jazmichaelking** [comment]
Just to add a use case here: I have multiple domains registered for a bilingual instance, eg toot.wales and toot.cymru, mastodon.wales and mastodon.cymru

Community members may wish to represent themselves as being more focused on Cymraeg (Welsh) than English. Being able to use member@toot.cymru instead of member@toot.wales would be of great benefit to them in how they want to represent themselves on the social web. 

Secondly, while I am personally a huge fan of "toot" as a branded moniker, I understand that some accounts may prefer a less frivolous-sounding handle for their activities. Being able to offer serious.domain and fun.domain as options would greatly improve my ability to market and advocate for the Fediverse generally. 

I am reminded of the proliferation of custom domain free email services that preceded everyone having an email account on one of five or so domains. At sign-up, I could choose from a list of domains available to me from the service. Those were fun services that allowed millions of people to express themselves with a domain that best suited them, without having to go through the cost and hassle of registering their own domain. There are fewer than 400 million domain names registered, for 5.5 billion Internet users. Not everyone has or wants their own domain.

I believe expanding and adding detail to the existing [ALTERNATE_DOMAINS](https://docs.joinmastodon.org/admin/config/#alternate_domains) functionality would be incredibly helpful for those of us who simply want to offer multiple domains (even if the login / account name is always tied to the primary domain). If I knew this to be a robust, functional, supported feature I would be offering multiple domains today. And - if it were to be detailed - I can see how this could lead to future conversations and enhancements that might allow for monetising additional custom/personal domains to help cover operating costs, as well as potentials for account portability, although I think both of those options are for a later discussion.

I leave it to more-informed commenters to ponder whether or not the user account would/should/could be discoverable by domains other than their chosen preference (I would hope not).

**Comment by michaelhofer** [comment]
Another use case where this might be useful is for further customisation and idividualisation of accounts. There are different projects and sites I support where a custom handle is wanted (similar to BlueSky's offering), but only a single account per domain is required (like myproject.com).

The solution with `ALTERNATE_DOMAINS` does not seem to help in this scenario, since users would still see the xyz@myserver.tld handle.

**Comment by Exagone313** [comment]
As registries are holding domain names for political reasons, it could be interesting to explore this topic again.
